### PR TITLE
feat: Authentication and Authorization (#40009)

### PR DESCRIPTION
**Pull request by ajayrose25**

### CODE REVIEW

File: src/ controllers/trainee/routes.ts [hunk 1]
```diff
@@ -2,13 +2,14 @@ import { Router } from 'express';
 import TraineeController from './Controller';
 import  { validationHandler }  from '../../libs/routes';
 import { validation } from './validation';
-// import {  authMiddleware }  from '../../libs/routes';
+import authMiddleware from '../../libs/routes/authMiddleware';
 import Controller from './Controller';
+import { permissions } from '../../libs/routes/constants';
 const traineeRouter = Router();
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/routes.ts]: define the type of TraineeRouter

File: src/config/configuration.ts [hunk 2]
```diff
@@ -9,4 +9,4 @@ const configuration: Iconfig = Object.freeze( {
     secretKey: process.env.SECRET_KEY
```
Comment (VisaaliSuccessive) [inline src/config/configuration.ts:9]: destructure this

File: src/libs/routes/authMiddleware.ts [hunk 3]
```diff
@@ -33,4 +33,4 @@ export default (module, permissionType) => (req: Request, res: Response, next: N
      message: error.message
    });
  }
-}; 
\ No newline at end of file
+};
```
Comment (VisaaliSuccessive) [inline src/libs/routes/authMiddleware.ts:48]: 1.  use destructuring in secretKey, role
2. verify in the case of no token 
3. give proper error messages saying "Unauthorized"
4. in the condition of hasPermission the message is not proper it should be "Permission denied"
5. pass the user in request user also 
6. call next function at the end

File: src/libs/routes/validationHandler.ts [hunk 4]
```diff
@@ -57,21 +50,21 @@ export default (validation) => (req: Request, res: Response, next: NextFunction)
                 const validateValues = values.filter(item => item);
                 console.log(validateValues, 'validateValues');
                 if (values.length !== validateValues.length)
-                    next({ message: `${key} is required` });
+                    next({ message: `${key} is invalid or it is not a number` });
```
Comment (VisaaliSuccessive) [inline src/libs/routes/validationHandler.ts]: give proper message

File: src/libs/routes/validationHandler.ts [hunk 5]
```diff
@@ -57,21 +50,21 @@ export default (validation) => (req: Request, res: Response, next: NextFunction)
...
             }
             if (keyData && keyData.default === 10 && keyData.number) {
                 console.log('____IN LIMIT____');
                 const validateValues = values.filter(item => item);
                 console.log(validateValues, 'validateValues');
                 if (values.length !== validateValues.length)
-                    next({ message: `${key} is required` });
+                    next({ message: `${key} is invalid or it is not a number` });
```
Comment (VisaaliSuccessive) [inline src/libs/routes/validationHandler.ts]: proper message

File: src/ controllers/trainee/routes.ts (hunk 1 above)
Comment (ajayrose25) [reply src/ controllers/trainee/routes.ts]: Done

File: src/config/configuration.ts (hunk 2 above)
Comment (ajayrose25) [reply src/config/configuration.ts:9]: Done

File: src/libs/routes/authMiddleware.ts (hunk 3 above)
Comment (ajayrose25) [reply src/libs/routes/authMiddleware.ts:48]: Done

File: src/libs/routes/validationHandler.ts (hunk 4 above)
Comment (ajayrose25) [reply src/libs/routes/validationHandler.ts]: Done

File: src/libs/routes/validationHandler.ts (hunk 5 above)
Comment (ajayrose25) [reply src/libs/routes/validationHandler.ts]: Done

File: src/libs/routes/authMiddleware.ts (hunk 3 above)
Comment (VisaaliSuccessive) [reply src/libs/routes/authMiddleware.ts:48]: 1,2,5 is not done

File: src/config/configuration.ts (hunk 2 above)
Comment (VisaaliSuccessive) [reply src/config/configuration.ts:9]: not done

**Review by VisaaliSuccessive — APPROVED** [review]

**Pull request merged**